replica/table.cc

```diff
@@ -2983,9 +2983,67 @@ bool table::requires_cleanup(const sstables::sstable_set& set) const {
     }));
 }
 
-future<> table::cleanup_tablet(locator::tablet_id) {
-    co_await flush();
-    // FIXME: Remove sstables
+future<> compaction_group::wipe() {
+    class compaction_group_cleaner : public row_cache::external_updater_impl {
+        table& _t;
+        compaction_group& _cg;
+        const lw_shared_ptr<sstables::sstable_set> _empty_main_set;
+        const lw_shared_ptr<sstables::sstable_set> _empty_maintenance_set;
+        std::vector<sstables::shared_sstable>& _all_sstables;
+        std::optional<compaction_manager::compaction_reenabler> _compaction_guard;
+    private:
+        lw_shared_ptr<sstables::sstable_set> empty_sstable_set() const {
+            return make_lw_shared<sstables::sstable_set>(_t._compaction_strategy.make_sstable_set(_t._schema));
+        }
+    public:
+        explicit compaction_group_cleaner(compaction_group& cg, std::vector<sstables::shared_sstable>& all_sstables)
+                : _t(cg._t)
+                , _cg(cg)
+                , _empty_main_set(empty_sstable_set())
+                , _empty_maintenance_set(empty_sstable_set())
+                , _all_sstables(all_sstables)
+        {
+        }
+        virtual future<> prepare() override {
+            _compaction_guard.emplace(co_await _t.get_compaction_manager().stop_and_disable_compaction(_cg.as_table_state()));
+            co_await _cg.flush();
+
+            // Capture sstables after flush, and with compaction disabled, to avoid missing any.
+            auto set = _cg.make_compound_sstable_set();
+            _all_sstables.reserve(set->size());
+            set->for_each_sstable([this] (const sstables::shared_sstable& sst) mutable {
+                _all_sstables.push_back(sst);
+            });
+        }
+        virtual void execute() override {
+            _cg.set_main_sstables(std::move(_empty_main_set));
+            _cg.set_maintenance_sstables(std::move(_empty_maintenance_set));
+        }
+    };
+
+    std::vector<sstables::shared_sstable> all_sstables;
+    auto updater = row_cache::external_updater(std::make_unique<compaction_group_cleaner>(*this, all_sstables));
+
+    auto p_range = to_partition_range(token_range());
+    tlogger.debug("Invalidating range {} for compaction group {} during cleanup.", p_range, group_id());
+    co_await _t._cache.invalidate(std::move(updater), p_range);
+
+    co_await delete_sstables_atomically(std::move(all_sstables));
+}
+
+future<> table::cleanup_tablet(locator::tablet_id tid) {
+    auto& cg_ptr = _compaction_groups[tid.value()];
+
+    if (!cg_ptr) {
+        throw std::runtime_error(format("Cannot cleanup tablet {} because it is not allocated in this shard", tid));
```

Comment:
To make tid meaningful, you need to also print table id. tablet_id is only unique within a table.

Reply:
will fix, thanks.